[FIX #3573] adding scripts/list-pull-requests.sh

jwagener opened the pull request:
Fixes #3573 

### Summary:

This script fetches the pull requests refs from upstream, then retrieves the log for the given commit range and outputs a link to the pull request for each given commit.

status: ready

jwagener committed:
[FIX #3573] adding scripts/list-pull-requests.sh and scripts/generate-pull-request-report.sh

Signed-off-by: Julien Eluard <julien.eluard@gmail.com>

ghost commented:
Hey @jwagener, thanks for making your first pull request in status-react! :heart:

jwagener commented:
Here is an example of the script usage. (GitHub will shorten the links, the real output contains the full url & commit hash). Happy to adjust it to your needs.

`$ ./scripts/list-pull-requests.sh 1248a469a6b26f92568be30f3c155d5e9708d7ca..HEAD

https://github.com/status-im/status-react/pull/3696 ef30913c8717e5ba68e086ad74b746c241501a79
https://github.com/status-im/status-react/pull/3673 e0f42d28d7b8061141ca1eee266d94cb93d5a055
https://github.com/status-im/status-react/pull/3688 e185acd92be61a8ab058d5b1494d5e51c7388341
https://github.com/status-im/status-react/pull/3664 c1c5854cb1c1204ae4f477eaafd8afdc8959d7df
https://github.com/status-im/status-react/pull/3653 f7f41d2f37e2163fe2c65676db56561747288e48
https://github.com/status-im/status-react/pull/3676 83b5312a05e1d5bd4e1ed8716a013828b3d765c1
https://github.com/status-im/status-react/pull/3674 7001b8c396d10ed420b3484935b54e99261c9b6f
https://github.com/status-im/status-react/pull/3663 8b34c5e728b61283c52737b15df30b66e82d32a1
https://github.com/status-im/status-react/pull/3592 b0a2bbe27b0aab7241b8ef42486bcc0d6539ebaa
~/Projects/Status/status-react [feature/list-pull-requests-#3573] $`

jeluard commented:
@jwagener Thanks !! I am wondering what would be the best output so that we could use PRs #s from others scripts. Any idea about that?

jwagener commented:
@jeluard Could you give me an example for that? Not sure how other scripts would play into that.

I've added an easy way to configure the output using the git log pretty format (https://git-scm.com/docs/pretty-formats).

Now it provides a bit more context (commit message, author):

`$ scripts/list-pull-requests.sh 1248a469a6b26f92568be30f3c155d5e9708d7ca..HEAD

https://github.com/status-im/status-react/pull/3696
ef30913c Johannes Wagener: [FIX #3573] adding scripts/list-pull-requests.sh
`

The only restriction is, that we unfortunately can't easily show the PR title, as it is not stored in git. This would require to add an API call to GitHub.

<blockquote><img src="/favicon.ico" width="48" align="right"><div><strong><a href="https://git-scm.com/docs/pretty-formats">Git - pretty-formats Documentation</a></strong></div></blockquote>

jeluard commented:
This script's output is mostly intended for computer consumption (not human).
One example of it's usage would be to produce some markdown file listing all PRs closed for a new release.

jwagener commented:
Ah ok!

In that case a tab separated one-line per pull request makes more sense. I'd propose the following:

(PR #) (tab) (GitHub API URL for PR) (tab) (list of relevant 
commit hashes separated by comma).

In case a PR wasn't squashed into on one commit, it would be good to group the commits by PR and include them as a list. I can add that.

The other script can then fetch the JSON for the PRs and generate the Markdown for the release. (Happy to look into that as well if requested)

jeluard commented:
Sounds perfect!

jwagener commented:
I've changed the output accordingly and added a scripts/generate-pull-request-report.sh.
For the later I've leaned a bit upon merge-pr.sh (using jq to parse json)
The template can be easily adjusted.

Example:

`$ ./scripts/list-pull-requests.sh c1c5854cb1c1204ae4f477eaafd8afdc8959d7df..HEAD  | ./scripts/generate-pull-request-report.sh`

# Pull Request Report

## Dont throw error on empty data (#3693)

URL:     https://github.com/status-im/status-react/pull/3693
Creator: cammellos
Labels:  Tested - OK


fixes #3590

### Summary:

This prevents the app crashing when resuming operations.

It was introduced on https://github.com/status-im/status-react/commit/ed7705ca7feb8186d6d48d8b485bb77c2faa3df4

It does not tackle the issue as I was not able to get the feature working on my device (clicking on settings does not redirects me to settings).

Pushing this as a quick fix, but needs to be investigated further.


status: ready


## [FIX #3573] adding scripts/list-pull-requests.sh (#3696)

URL:     https://github.com/status-im/status-react/pull/3696
Creator: jwagener
Labels:  bounty, developer-xp

Fixes #3573

### Summary:
...

jeluard commented:
Wow awesome work!

jeluard commented on scripts/generate-pull-request-report.sh:
```diff
@@ -0,0 +1,82 @@
+# Github Client Credentials
+# Required to increase rate limits to a useable level.
+
+GITHUB_CLIENT_ID=cae8611e191547e163bb
```
We probably want to externalize that?

jwagener replied:
Agree that it is a bit weird, but since it's only used for the increased rate limit (5000 requests vs 60) and nothing else otherwise we could leave it for the sake of simplicity. 

If not, where do we move this?

jeluard replied:
Makes sense, let's keep it this way.

dmitryn commented on scripts/list-pull-requests.sh:
```diff
@@ -0,0 +1,46 @@
+# Fetches pull request refs and prints commits
+# for the given range grouped by pull request
+
+UPSTREAM_REMOTE="upstream"
```
Just a note, i don't have "upstream", but have "origin" because i'm using status-im/status-react repository directly. So i end up changing this line to `UPSTREAM_REMOTE="origin"`
to make it work.

jeluard replied:
@jwagener Any idea how we could handle both cases?

jwagener replied:
I changed it to use a separate remote name to fetch the PR refs. Now it should be completely independent from the developer remote setup.

jeluard commented:
@jwagener Awesome! Could you squash your commits?

jwagener commented:
Done. That was fun :)
